File: js/y.js
var Y = function (h) {
    return (function (f) {
        return f(f);
    })(function (f) {
        return h(function () {
            return f(f).apply(null, arguments);
        });
    });
};

var fact = Y(function (recur) {
    return function (n) {
        if (n < 2) return 1;
        return n * recur(n - 1);
   };
})

console.log(fact(5));

File: ruby/y.rb
Y = lambda {|h|
  lambda {|f|
    f[f]
  }[lambda {|f|
    h[lambda {|*args| f[f][*args] }]
  }]
}

fact = lambda {|recur|
  lambda {|n|
    return 1 if n < 2
    n * recur[n - 1]
  }
}

puts Y[fact][5]
